lib/auth/oidc.go

```diff
@@ -36,38 +37,103 @@ import (
 	"github.com/gravitational/trace"
 )
 
+func (s *AuthServer) getOrCreateOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
+	client, err := s.getOIDCClient(conn)
+	if err == nil {
+		return client, nil
+	}
+	if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	return s.createOIDCClient(conn)
+}
+
 func (s *AuthServer) getOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	config := oidc.ClientConfig{
-		RedirectURL: conn.GetRedirectURL(),
-		Credentials: oidc.ClientCredentials{
-			ID:     conn.GetClientID(),
-			Secret: conn.GetClientSecret(),
-		},
-		// open id notifies provider that we are using OIDC scopes
-		Scope: utils.Deduplicate(append([]string{"openid", "email"}, conn.GetScope()...)),
+	clientPack, ok := s.oidcClients[conn.GetName()]
+	if !ok {
+		return nil, trace.NotFound("connector %v is not found", conn.GetName())
 	}
 
-	clientPack, ok := s.oidcClients[conn.GetName()]
+	config := oidcConfig(conn)
 	if ok && oidcConfigsEqual(clientPack.config, config) {
 		return clientPack.client, nil
 	}
+
 	delete(s.oidcClients, conn.GetName())
+	return nil, trace.NotFound("connector %v has updated the configuration and is invalidated", conn.GetName())
 
+}
+
+func (s *AuthServer) createOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
+	config := oidcConfig(conn)
 	client, err := oidc.NewClient(config)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	client.SyncProviderConfig(conn.GetIssuerURL())
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.WebHeadersTimeout)
+	defer cancel()
+
+	go func() {
+		defer cancel()
+		client.SyncProviderConfig(conn.GetIssuerURL())
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
Select does not support fall through, so this case statement does nothing.

Reply:
what do you mean? this select is waiting until the context gets cancelled in this case it will continue executing on line 90, or in case if the auth server is shutting down,  returns the error.

Reply:
Got it, read it wrong.